switchtec.c

```diff
@@ -568,7 +569,8 @@ static ssize_t switchtec_dev_read(struct file *filp, char __user *data,
 out:
 	mutex_unlock(&stdev->mrpc_mutex);
 
-	if (stuser->status == SWITCHTEC_MRPC_STATUS_DONE)
+	if (stuser->status == SWITCHTEC_MRPC_STATUS_DONE ||
+	    stuser->status == SWITCHTEC_MRPC_STATUS_ERROR)
```

Comment:
Hmm? Shouldn't we still return an error code, not the size?

Reply:
In this SWITCHTEC_MRPC_STATUS_ERROR case, the user code should take the MRPC execution as finished, and check return code (which must be non-zero to indicate to error). And I was suggesting firmware to return SWITCHTEC_MRPC_STATUS_DONE in this case, since we must check the return code (which indicate the specific error) no matter what the status value is ERROR or DONE (Unfortunately, it's been there in the firmware for a while and released ). 

Therefore, the user space handling should be exactly the same as SWITCHTEC_MRPC_STATUS_DONE. The only difference is the return code could be zero or not in the SWITCHTEC_MRPC_STATUS_DONE case, and must be non-zero in the SWITCHTEC_MRPC_STATUS_ERROR  case. 

Below is the current user code to handle MRPC completion. With this patch, we can keep it as is, since it has already handles the process properly.

```
static int read_resp(struct switchtec_linux *ldev, void *resp,
                     size_t resp_len)
{
        int32_t ret;
        size_t bufsize = sizeof(uint32_t) + resp_len;
        char buf[bufsize];

        ret = read(ldev->fd, buf, bufsize);

        if (ret < 0)
                return ret;

        if (ret != bufsize) {
                errno = EIO;
                return -errno;
        }

        memcpy(&ret, buf, sizeof(ret));
        if (ret)
                errno = ret;

        if (!resp)
                return ret;

        memcpy(resp, &buf[sizeof(ret)], resp_len);

        return ret;
```

Reply:
Ok, seems sensible.